crates/admin/src/schema_registry/updater.rs

```diff
@@ -260,6 +257,169 @@ impl SchemaUpdater {
         Ok(deployment_id)
     }
 
+    pub fn update_deployment(
+        &mut self,
+        deployment_id: DeploymentId,
+        deployment_metadata: DeploymentMetadata,
+        services: Vec<endpoint_manifest::Service>,
+    ) -> Result<(), SchemaError> {
+        let proposed_services: HashMap<_, _> = services
+            .into_iter()
+            .map(|c| ServiceName::try_from(c.name.to_string()).map(|name| (name, c)))
+            .collect::<Result<HashMap<_, _>, _>>()?;
+
+        // Look for an existing deployment with this ID
+        let Some((_, existing_deployment)) = self
+            .schema_information
+            .find_existing_deployment_by_id(&deployment_id)
+        else {
+            return Err(SchemaError::NotFound(format!(
+                "deployment with id '{deployment_id}'"
+            )));
+        };
+
+        let mut services_to_remove = Vec::default();
+
+        for service in &existing_deployment.services {
+            if !proposed_services.contains_key(&service.name) {
+                services_to_remove.push(service.name.clone());
+            }
+        }
+
+        if !services_to_remove.is_empty() {
+            // we don't allow removing services as part of update deployment
+            return Err(SchemaError::Deployment(DeploymentError::RemovedServices(
+                services_to_remove,
+            )));
+        }
+
+        let mut services_to_add = HashMap::with_capacity(proposed_services.len());
+
+        // Compute service schemas
+        for (service_name, service) in proposed_services {
+            let service_type = ServiceType::from(service.ty);
+            let handlers = DiscoveredHandlerMetadata::compute_handlers(
+                service
+                    .handlers
+                    .into_iter()
+                    .map(|h| {
+                        DiscoveredHandlerMetadata::from_schema(
+                            service_name.as_ref(),
+                            service_type,
+                            h,
+                        )
+                    })
+                    .collect::<Result<Vec<_>, _>>()?,
+            );
```

Comment:
I think you're missing somewhere to preserve the private handler flag? Maybe add a unit test to check that.